Do not leak cursors

davidegrohmann opened the pull request:
This PR replaces #9184

davidegrohmann committed:
Do not leak Node-Relationship Cursors in Core API

Make sure that inner cursor is closed when the user called close on
the public ResourceIterator

davidegrohmann committed:
Do not leak cursors in interpreted runtime

This commit does _not_ fix the compiled runtime

davidegrohmann committed:
No monkeys

systay commented on enterprise/cypher/acceptance-spec-suite/src/test/scala/org/neo4j/internal/cypher/acceptance/AggregationAcceptanceTest.scala:
```diff
@@ -23,12 +23,6 @@ import org.neo4j.cypher.{ExecutionEngineFunSuite, NewPlannerTestSupport}
 
 class AggregationAcceptanceTest extends ExecutionEngineFunSuite with NewPlannerTestSupport {
 
-  test("apa") {
```
Oops :)

davidegrohmann committed:
Add tests for avoiding leaks in cypher runtime

systay commented:
This change feels much cleaner and easier to understand than the ThreadLocal variation in #9184. 👍 

There are no unit tests for the new ResourceManager, but other than that, LGTM

davidegrohmann committed:
Add tests for avoiding leaks in Core API

davidegrohmann commented:
@systay I have added the tests now

pontusmelke commented on community/cypher/cypher-compiler-3.2/src/test/scala/org/neo4j/cypher/internal/compiler/v3_2/spi/QueryContextAdaptation.scala:
```diff
@@ -32,6 +32,8 @@ trait QueryContextAdaptation {
 
   override type EntityAccessor = this.type
 
+  override def resources: CloseableResource = ???
```
is this safe to leave unimplemented? If so maybe a comment to future readers :)

Oh I see we do a lot of that already, never mind then...

davidegrohmann replied:
Basically almost all the methods in that class are unimplemented, moreover that class is used only in tests. I don't think there is any need to add any explanation.

pontusmelke replied:
👍

thobe commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/ResourceManager.scala:
```diff
@@ -0,0 +1,50 @@
+    val iterator = resources.iterator()
+    while ( iterator.hasNext )
+    {
+      iterator.next().close()
```
Should we be handling potential exceptions from the `close()` method?

davidegrohmann replied:
Yep, you are right we should.

thobe commented on community/cypher/cypher/src/test/scala/org/neo4j/cypher/internal/spi/BeansAPIRelationshipIteratorTest.scala:
```diff
@@ -0,0 +1,55 @@
+    iterator.close()
+    verify(relationships).close()
+    verifyNoMoreInteractions(relationships)
+  }
```
How about a test case to ensure that `close()` is idempotent?

davidegrohmann replied:
Good point, I'll add one.

thobe reviewed (commented):
Looks like a nice approach to solving the resource leak hazard.

davidegrohmann committed:
Fix close all resources on close even when exceptions are thrown

ResourceManager will handle exceptions when closing internal resources
by making sure all resources are closed before throwing an exception
at the end. If more than one exception is thrown internally, the first
one caught will be rethrown and the other are added to such exceptions
as suppressed exceptions.

systay commented:
To be merged straight after release of Neo4j 3.2
